cypress-io/cypress
# chore: prep npm packages for use with Cypress v10

- Closes https://github.com/cypress-io/cypress/issues/21944

### User facing changelog
Publish Cypress v10 compatible npm packages

### Additional details
Removes `"private": true` from the npm packages that we had disabled and fixes a few bugs in our implementation that would have caused issues. Notable changes:

- Replacing the require path for `@cypress/webpack-batteries-included-preprocessor` in webpack-dev-server from `__dirname` to `cypressBinaryRoot`. Relying on `__dirname` only works when sourcing from within the binary (or sourcing from the packages installed node_modules).
- Revert `find-up` to 5.0.0 as >6.0.0 is ESM only. We were relying on `local-pkg` to import it which uses `require`. Removed both packages.
- Restoed webpack-dev-server changelog that got removed in a previous merge
- Fixed types for when using the dev-server function signature

The scope for the original issue became very bloated so I'm not ticking all the boxes from that issue (such as linting non-used deps or releasing as `RC`). Not sure why we came to the decision to publish them as RC but I thoroughly tested all the packages locally and I'm confident we can go forward with a normal breaking change.

I ran the `npm-release.js` script locally and the new versions are as expected:
```
node ./scripts/npm-release.js 
Found the following public packages: create-cypress-tests, @cypress/schematic, @cypress/eslint-plugin-dev, @cypress/mount-utils, @cypress/react, @cypress/vite-dev-server, @cypress/vue, @cypress/vue2, @cypress/webpack-batteries-included-preprocessor, @cypress/webpack-dev-server, @cypress/webpack-preprocessor

Finding package versions...

Cypress binary: 10.0.3

create-cypress-tests
Current version: 1.3.0
Next version: 2.0.0

@cypress/schematic
Current version: 2.0.0
Next version: N/A

@cypress/eslint-plugin-dev
Current version: 5.3.0
Next version: N/A

@cypress/mount-utils
Current version: 1.0.2
Next version: 2.0.0

@cypress/react
Current version: 5.12.5
Next version: 6.0.0

@cypress/vite-dev-server
Current version: 2.2.3
Next version: 3.0.0

@cypress/vue
Current version: 3.1.2
Next version: 4.0.0

@cypress/vue2
Current version: N/A
Next version: 1.0.0

@cypress/webpack-batteries-included-preprocessor
Current version: 2.2.3
Next version: N/A

@cypress/webpack-dev-server
Current version: 1.8.4
Next version: 2.0.0

@cypress/webpack-preprocessor
Current version: 5.12.0
Next version: N/A

Found a new release for the following packages: create-cypress-tests, @cypress/mount-utils, @cypress/react, @cypress/vite-dev-server, @cypress/vue, @cypress/vue2, @cypress/webpack-dev-server
```

We should merge with care. I included two commits, one with all the changes to the files in `npm/` with the `BREAKING CHANGE` commit message and another with the rest of the changes. I didn't want to include a `BREAKING CHANGE` commit message that touched files outside of npm. When merging, I'd like to not squash.

Need to follow up with DX about documentation. There is a [section](https://docs.cypress.io/guides/component-testing/component-framework-configuration#Custom-Dev-Server) about custom-dev-servers, gotta check if this needs to be expanded.

### Steps to test
I tested the packages by running `yarn pack` and installing the generated `.tgz`. I added a few system tests for the dev-servers to verify the use of the function signature.

### How has the user experience changed?
na

### PR Tasks
